src/shared/codelens/pythonCodeLensProvider.ts

```diff
@@ -0,0 +1,109 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+import * as path from 'path'
+import * as vscode from 'vscode'
+
+import { DebugConfiguration } from '../../lambda/local/debugConfiguration'
+import { LambdaHandlerCandidate } from '../lambdaHandlerSearch'
+import { DefaultSamCliProcessInvoker, DefaultSamCliTaskInvoker, } from '../sam/cli/samCliInvoker'
+
+import {
+    CodeLensProviderParams,
+    getInvokeCmdKey,
+    getLambdaHandlerCandidates,
+    getLogger,
+    makeCodeLenses,
+    OutputChannelName
+} from './codeLensUtils'
+import { LambdaLocalInvokeArguments, LocalLambdaRunner } from './localLambdaRunner'
+
+const logger = getLogger(OutputChannelName.ToolKit)
+
+export const PYTHON_LANGUAGE = 'python'
+// export const PYTHON: vscode.DocumentFilter[] = [ // TODO: Use this or PYTHON_ALLFILES?
+//     { scheme: 'file', language: PYTHON_LANGUAGE },
+//     { scheme: 'untitled', language: PYTHON_LANGUAGE }
+// ]
+export const PYTHON_ALLFILES = [
+    { language: PYTHON_LANGUAGE }
+]
+
+// TODO: Fix this! Implement a more robust/flexible solution. This is just a basic minimal proof of concept.
```

Comment:
Try using `detectLocalTemplates` from `src/lambda/local/detectLocalTemplates.ts`. You can get the workspace folder containing `filepath` by using `vscode.workspace.getWorkspaceFolder(vscode.Uri.file(filepath)).

Reply:
Hm... I don't know how relevant the VS Code workspace is. A user could have one big VS Code workspace with a hundred SAM project folders in it. It seem the most reasonable approach is to go up form the current file, one directory at a time keeping track of the current and last directory. Once we encounter a YAML template, then the SAM project directory would be the last directory encountered before the one containing the YAML file. Anyway, implementing that is for the next issue. 

If `requirements.txt` was found, we could use the containing directory as the SAM project folder.

This currently works to support Python in a very specific case. The purpose of this CR is to display the code lens, not to implement all of the associated Python functionality. There are a few options here:

1. Approve it and do the Python work in another CR
2. Keep this open and continue Python work, engulfing the other associated CRs
3. Comment out the PythonCode lens to restore in a later CR once support is in order. e.g.

```
async function activateCodeLensProviders(
    configuration: SettingsConfiguration
): Promise<vscode.Disposable[]> {
    const disposables: vscode.Disposable[] = []
    const providerParams = {
        configuration,
        toolkitOutputChannel
    }

    tsLensProvider.initialize(providerParams)
    disposables.push(vscode.languages.registerCodeLensProvider(
        [
            {
                language: 'javascript',
                scheme: 'file',
            }
        ],
        tsLensProvider.makeTypescriptCodeLensProvider()
    ))

    // pyLensProvider.initialize(providerParams)
    // disposables.push(vscode.languages.registerCodeLensProvider(
    //     pyLensProvider.PYTHON_ALLFILES,
    //     pyLensProvider.makePythonCodeLensProvider()
    // ))

    return disposables
}
```

Reply:
We should do 3 at this point, and the follow-up task #291 would set up the CodeLenses with proper python functionality, answering all Python related TODO's earmarked in this PR.